compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/types/TypeBuilder.java

```diff
@@ -0,0 +1,389 @@
+import org.wso2.ballerinalang.compiler.semantics.model.types.BReadonlyType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BRecordType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BServiceType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BStreamType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BStructureType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTableType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTupleType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTypeVisitor;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BTypedescType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BUnionType;
+import org.wso2.ballerinalang.compiler.semantics.model.types.BXMLType;
+import org.wso2.ballerinalang.compiler.util.TypeTags;
+import org.wso2.ballerinalang.util.Flags;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static io.ballerina.compiler.api.types.ParameterKind.DEFAULTABLE;
+import static io.ballerina.compiler.api.types.ParameterKind.REQUIRED;
+import static io.ballerina.compiler.api.types.ParameterKind.REST;
+import static java.util.Collections.unmodifiableList;
+import static java.util.stream.Collectors.toList;
+import static org.ballerinalang.model.types.TypeKind.OBJECT;
+import static org.ballerinalang.model.types.TypeKind.RECORD;
+
+/**
+ * A type visitor to build the public type for a given type.
+ *
+ * @since 2.0.0
+ */
+public class TypeBuilder implements BTypeVisitor<BType, BallerinaTypeDescriptor> {
+
+    private ModuleID moduleID;
+    private final Types types;
+    private final LangLibrary langLibrary;
+    private final Map<BType, BallerinaTypeDescriptor> typeCache;
+
+    public TypeBuilder(Types types, LangLibrary langLibrary) {
+        this.types = types;
+        this.typeCache = new HashMap<>();
+        this.langLibrary = langLibrary;
+    }
+
+    public BallerinaTypeDescriptor build(BType internalType) {
+        if (internalType == null || internalType.tag == TypeTags.NONE) {
+            return null;
+        }
+
+        if (typeCache.containsKey(internalType)) {
+            return typeCache.get(internalType);
+        }
+
+        moduleID = internalType.tsymbol == null ? null : new BallerinaModuleID(internalType.tsymbol.pkgID);
+        BallerinaTypeDescriptor publicType = internalType.accept(this, null);
+
+        if (publicType != null) {
+            List<LangLibMethod> langLibMethods = langLibrary.getMethods(publicType.kind());
+            langLibMethods = filterMethods(langLibMethods, internalType);
+            ((AbstractTypeDescriptor) publicType).setLangLibMethods(unmodifiableList(langLibMethods));
```

Comment:
Need this cast to `AbstractTypeDescriptor`?

Reply:
It's because the setter is not part of the public API.